### PR TITLE
Fix indexes on the Outbox table on SQL Server

### LINKED ISSUE #66: Outbox cleanup does full table scan on SQL Server
### What is the problem?

Currenly the Outbox on SQL Server creates two separate indexes on `Dispatched` and `DispatchedAt` columns.

Execution plan of the cleanup query:

```sql
delete top (@BatchSize) from table
where Dispatched = 'true'
    and DispatchedAt < @DispatchedBefore"
```

Indicates that a table scan is performed. All the other queries only use the primary key index (unique lookup by the primary key).

These two indexes are useless and have negative performance impact (they have to be updated on insert and when marking an Outbox record as dispatched).

### How to solve it?

A single filtered index will resolve it:

```sql
create index table_dispatched_idx on table (DispatchedAt) where Dispatched = 1
```

### PR DESCRIPTION
**Pull request by MarcinHoppe**
Fixes #66

### CODE REVIEW

**Comment by MarcinHoppe** [comment]
@Particular/sqlpersistence-maintainers The build is green and this is ready for review.

File: src/ScriptBuilder/Outbox/Create_MsSqlServer.sql [hunk 1]
```diff
@@ -32,23 +32,6 @@ begin
...
   exec(@createDispatchedAtIndex);
-end
-
-
-if not exists
-(
-    select *
-    from sys.indexes
-    where
-        name = 'Index_Dispatched' and
```
Comment (SimonCropp) [inline src/ScriptBuilder/Outbox/Create_MsSqlServer.sql]: if Index_Dispatched is no longer required should it be cleaned up?

Comment (MarcinHoppe) [inline src/ScriptBuilder/Outbox/Create_MsSqlServer.sql]: It should. Sounds like something that should be addressed in the upgrade guide?

**Comment by SzymonPobiega** [comment]
Apart from Simon's comments, LGTM.

File: src/ScriptBuilder/Outbox/Create_MsSqlServer.sql (hunk 1 above)
Comment (SimonCropp) [inline src/ScriptBuilder/Outbox/Create_MsSqlServer.sql]: fair enough. do u want to handle that? if not i can.

Comment (MarcinHoppe) [inline src/ScriptBuilder/Outbox/Create_MsSqlServer.sql]: Go on. I'll be happy to review.

Comment (SimonCropp) [inline src/ScriptBuilder/Outbox/Create_MsSqlServer.sql]: ok but it wont happen till monday

Comment (MarcinHoppe) [inline src/ScriptBuilder/Outbox/Create_MsSqlServer.sql]: Sure, I wouldn't have worked on it any sooner.

Comment (SimonCropp) [inline src/ScriptBuilder/Outbox/Create_MsSqlServer.sql]: actually this index was added in v2. so no doco needed

**Comment by SimonCropp** [comment]
@MarcinHoppe thanks for the effort of research and PR

**Comment by SimonCropp** [comment]
@DavidBoike sorry given u added those indexes i would appreciate your review of this change

File: src/ScriptBuilder/Outbox/Create_MsSqlServer.sql [hunk 2]
```diff
@@ -32,23 +32,6 @@ begin
   declare @createDispatchedAtIndex nvarchar(max);
   set @createDispatchedAtIndex = '
   create index Index_DispatchedAt
-  on ' + @tableName + '(DispatchedAt);';
+  on ' + @tableName + '(DispatchedAt) where Dispatched = 1;';
```
Comment (DavidBoike) [inline src/ScriptBuilder/Outbox/Create_MsSqlServer.sql]: So, this is a [filtered index](https://docs.microsoft.com/en-us/sql/relational-databases/indexes/create-filtered-indexes), which was new in SQL 2016, and we [currently support down to SQL 2012](https://docs.particular.net/nservicebus/sql-persistence/?version=sqlpersistence_2#supported-sql-implementations). :frown:

Comment (DavidBoike) [inline src/ScriptBuilder/Outbox/Create_MsSqlServer.sql]: Ok nevermind, I guess it was [introduced in SQL 2008](https://www.mssqltips.com/sqlservertip/1785/sql-server-filtered-indexes-what-they-are-how-to-use-and-performance-advantages/) and the Microsoft link with its **SQL Server (starting with 2016)** label at the top is full of it.

**Comment by DavidBoike** [comment]
I'm retroactively :+1: - really good use of a filtered index, glad we could take advantage.

**Comment by SimonCropp** [comment]
@DavidBoike thanks for double checking this

**Pull request merged**